implementations/rust/ockam/ockam/src/system/hooks/delivery.rs

```diff
@@ -0,0 +1,170 @@
+use crate::{Context, DelayedEvent, OckamError, OckamMessage, Result, Routed, SystemHandler};
+use ockam_core::{
+    async_trait,
+    compat::{boxed::Box, collections::BTreeMap, string::String, vec::Vec},
+    Address, Any, Decodable, Encodable,
+};
+
+#[derive(Clone, Default)]
+pub struct SenderConfirm {
+    next: Option<Address>,
+    journal: BTreeMap<Address, OckamMessage>,
+}
+
+fn to_str(v: &[u8]) -> &str {
+    core::str::from_utf8(v).unwrap()
+}
+
+#[async_trait]
+impl SystemHandler<Context, OckamMessage> for SenderConfirm {
+    async fn initialize(
+        &mut self,
+        ctx: &mut Context,
+        routes: &mut BTreeMap<String, Address>,
+    ) -> Result<()> {
+        self.next = Some(
+            routes
+                .remove("default")
+                .ok_or(OckamError::SystemInvalidConfiguration)?,
+        );
+        Ok(())
+    }
+
+    // The sender confirm hook stores a reference for the sent-message
+    // and registers a delayed event to re-check whether we received a
+    // confirmation or not.
+    async fn handle_message(
+        &mut self,
+        self_addr: Address,
+        ctx: &mut Context,
+        msg: Routed<OckamMessage>,
+    ) -> Result<()> {
+        trace!("SenderDelivery '{}' handling incoming message", self_addr);
+
+        match msg
+            .generic
+            .as_ref()
+            .and_then(|data| data.get("ockam.pipe.type"))
+            .map(|tt| to_str(tt))
+        {
+            // For a message with no type we register a delayed event
+            // and forward it to the next step in the system
+            None => {
+                let ack_id = Address::random(0);
+                let inner_msg = msg.body();
+                let outer_msg = OckamMessage::wrap(inner_msg)?
+                    .scope_data(ack_id.encode()?)
+                    .scope_data(self_addr.encode()?);
+                self.journal.insert(ack_id, outer_msg.clone());
+
+                // Register a delayed event to check whether we
+                // received an ACK for this message.  TODO: replace
+                // this with ockam_core::Heartbeat?
+                DelayedEvent::new(ctx, vec![self_addr.clone()].into(), {
+                    OckamMessage::new(Any)?.generic_data(
+                        "ockam.pipe.type",
+                        "ockam.pipe.resend_notify".as_bytes().to_vec(),
+                    )
+                })
+                .await?
+                .with_seconds(5)
+                .spawn();
+
+                // Forward the new message to the next address
+                ctx.send(self.next.as_ref().unwrap().clone(), outer_msg)
```

Comment:
This unwrap shouldn't trigger unless this is called prior to initialize (which is impossible[^1]), right?

[^1]: Short of someone manually calling it, which we don't/shouldn't support.

Reply:
Correct. `self.next` is set during worker initialisation. If someone were to create a worker instance and then manually call handle_message on it (with the correct message type and wrapping) _then_ it would fail